core/primitives/src/epoch_manager.rs

```diff
@@ -620,42 +618,158 @@ pub struct ValidatorWeight(ValidatorId, u64);
     use crate::version::PROTOCOL_VERSION;
     use borsh::{BorshDeserialize, BorshSerialize};
     use near_primitives_core::hash::CryptoHash;
     use near_primitives_core::types::{
-        AccountId, Balance, EpochHeight, ProtocolVersion, ValidatorId,
+        AccountId, Balance, EpochHeight, NumBlocks, ProtocolVersion, ValidatorId,
     };
     use rand::SeedableRng;
     use rand_chacha::ChaCha20Rng;
     use smart_default::SmartDefault;
     use std::collections::{BTreeMap, HashMap};
 
-    use crate::types::validator_stake::ValidatorStakeV1;
     use crate::{epoch_manager::RngSeed, rand::WeightedIndex};
     use near_primitives_core::{
         checked_feature,
         hash::hash,
         types::{BlockHeight, ShardId},
     };
 
-    pub use super::EpochInfoV1;
-
     /// Information per epoch.
+    /// DEPRECATED. Remove after DB version 40 and protocol feature
+    /// StatelessValidationV0 are released.
+    #[derive(
+        SmartDefault,
+        BorshSerialize,
+        BorshDeserialize,
+        Clone,
+        Debug,
+        PartialEq,
+        Eq,
+        serde::Serialize,
+    )]
+    pub struct EpochInfoV1 {
+        /// Ordinal of given epoch from genesis.
+        /// There can be multiple epochs with the same ordinal in case of long forks.
+        pub epoch_height: EpochHeight,
+        /// List of current validators.
+        pub validators: Vec<ValidatorStakeV1>,
+        /// Validator account id to index in proposals.
+        pub validator_to_index: HashMap<AccountId, ValidatorId>,
+        /// Settlement of validators responsible for block production.
+        pub block_producers_settlement: Vec<ValidatorId>,
+        /// Per each shard, settlement validators that are responsible.
+        pub chunk_producers_settlement: Vec<Vec<ValidatorId>>,
+        /// Settlement of hidden validators with weights used to determine how many shards they will validate.
+        pub hidden_validators_settlement: Vec<ValidatorWeight>,
+        /// List of current fishermen.
+        pub fishermen: Vec<ValidatorStakeV1>,
+        /// Fisherman account id to index of proposal.
+        pub fishermen_to_index: HashMap<AccountId, ValidatorId>,
+        /// New stake for validators.
+        pub stake_change: BTreeMap<AccountId, Balance>,
+        /// Validator reward for the epoch.
+        pub validator_reward: HashMap<AccountId, Balance>,
+        /// Validators who are kicked out in this epoch.
+        pub validator_kickout: HashMap<AccountId, LegacyValidatorKickoutReasonV39>,
+        /// Total minted tokens in the epoch.
+        pub minted_amount: Balance,
+        /// Seat price of this epoch.
+        pub seat_price: Balance,
+        /// Current protocol version during this epoch.
+        pub protocol_version: ProtocolVersion,
+    }
+
     #[derive(BorshSerialize, BorshDeserialize, Clone, Debug, PartialEq, Eq, serde::Serialize)]
-    pub enum EpochInfo {
+    /// DEPRECATED. Remove after DB version 40 and protocol feature
```

Comment:
how will this not impact the backwards compatibility of the code? is this implementation specific?

Reply:
I don't get the question, could you be more specific? LegacyEpochInfo has the same serialization as current EpochInfo; EpochInfo::V4 can still be read for statelessnet nodes. Implementation of EpochInfo functions doesn't change.

There is a non-trivial thing that before some logic was depending on which version of enum we are using; but in fact, it is enough to do a protocol version check in the same enum.

Reply:
Before this change takes an effect, what will be the struct that the previous protocol version will read. I think I am lost in the sequence of events leading to this change. Is there a write-up of the whole sequence of events and why did we end up with this change?

Reply:
The structure read is independent from the protocol version. It depends on DB_VERSION used.

For DB_VERSION < 40, nodes are using LegacyEpochInfo (here) which supports all V1-V4 versions.
For DB_VERSION = 40, all LegacyEpochInfos will be replaced with EpochInfo which supports only V4. No protocol changes are introduced, the logic based on specific protocol version remains the same.

The event leading to that is the accidental change of `ValidatorKickoutReason`. The change is good on its own, but maintaining V1-V3 would mean migrating V1-V3 separately. I don't want to do this, because in fact, we never needed V1-V3, V4 is enough to support all protocol versions. So I move everything to V4.